src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -83,6 +83,21 @@ class SimpleCanonicalResult {
 // does not exist one such number, the calculation will overflow and return a
 // value < lo.
 //
+// Formally, this function tries to find the minimum value that is not less
+// than lo and satisfies bits, assuming such value exists. The cases where such
+// value does not exists automatically follows.
+//
+// If there exists a value not less than lo and satisfies bits, then this
+// function will always find a such value. The conversion is also true, that is
+// if this function finds a value not less than lo and satisfies bits, then it
+// must trivially be the case that there exists a such value. As a result, the
+// negation of those statements are also equivalent, there does not exists a
+// value not less than lo and satisfies bits if and only if this function does
+// not return a such value.
+//
+// In practice, since the algorithm always ensures that the returned value
+// satisfies bits, we only need to check if it is not less than lo.
```

Comment:
Ah nice. Ok. So now we just have to prove that the result satisfies bits in all cases :)

Reply:
I'll check the proofs again to see if this is clear enough later.